File: mem0/vector_stores/elasticsearch.py
import logging
from typing import Any, Dict, List, Optional

try:
    from elasticsearch import Elasticsearch
    from elasticsearch.helpers import bulk
except ImportError:
    raise ImportError(
        "Elasticsearch requires extra dependencies. Install with `pip install elasticsearch`"
    ) from None

from pydantic import BaseModel

from mem0.configs.vector_stores.elasticsearch import ElasticsearchConfig
from mem0.vector_stores.base import VectorStoreBase

logger = logging.getLogger(__name__)


class OutputData(BaseModel):
    id: str
    score: float
    payload: Dict


class ElasticsearchDB(VectorStoreBase):
    def __init__(self, **kwargs):
        config = ElasticsearchConfig(**kwargs)

        # Initialize Elasticsearch client
        if config.cloud_id:
            self.client = Elasticsearch(
                cloud_id=config.cloud_id,
                api_key=config.api_key,
                verify_certs=config.verify_certs,
            )
        else:
            self.client = Elasticsearch(
                hosts=[f"{config.host}" if config.port is None else f"{config.host}:{config.port}"],
                basic_auth=(config.user, config.password) if (config.user and config.password) else None,
                verify_certs=config.verify_certs,
            )

        self.collection_name = config.collection_name
        self.vector_dim = config.embedding_model_dims

        # Create index only if auto_create_index is True
        if config.auto_create_index:
            self.create_index()

    def create_index(self) -> None:
        """Create Elasticsearch index with proper mappings if it doesn't exist"""
        index_settings = {
            "mappings": {
                "properties": {
                    "text": {"type": "text"},
                    "embedding": {
                        "type": "dense_vector",
                        "dims": self.vector_dim,
                        "index": True,
                        "similarity": "cosine",
                    },
                    "metadata": {"type": "object"},
                    "user_id": {"type": "keyword"},
                    "hash": {"type": "keyword"},
                }
            }
        }

        if not self.client.indices.exists(index=self.collection_name):
            self.client.indices.create(index=self.collection_name, body=index_settings)
            logger.info(f"Created index {self.collection_name}")
        else:
            logger.info(f"Index {self.collection_name} already exists")

    def create_col(self, name: str, vector_size: int, distance: str = "cosine") -> None:
        """Create a new collection (index in Elasticsearch)."""
        index_settings = {
            "mappings": {
                "properties": {
                    "vector": {"type": "dense_vector", "dims": vector_size, "index": True, "similarity": "cosine"},
                    "payload": {"type": "object"},
                    "id": {"type": "keyword"},
                }
            }
        }

        if not self.client.indices.exists(index=name):
            self.client.indices.create(index=name, body=index_settings)
            logger.info(f"Created index {name}")

    def insert(
        self, vectors: List[List[float]], payloads: Optional[List[Dict]] = None, ids: Optional[List[str]] = None
    ) -> List[OutputData]:
        """Insert vectors into the index."""
        if not ids:
            ids = [str(i) for i in range(len(vectors))]

        if payloads is None:
            payloads = [{} for _ in range(len(vectors))]

        actions = []
        for i, (vec, id_) in enumerate(zip(vectors, ids)):
            action = {"_index": self.collection_name, "_id": id_, "vector": vec, "payload": payloads[i]}
            actions.append(action)

        bulk(self.client, actions)

        # Return OutputData objects for inserted documents
        results = []
        for i, id_ in enumerate(ids):
            results.append(
                OutputData(
                    id=id_,
                    score=1.0,  # Default score for inserts
                    payload=payloads[i],
                )
            )
        return results

    def search(self, query: List[float], limit: int = 5, filters: Optional[Dict] = None) -> List[OutputData]:
        """Search for similar vectors using KNN search with pre-filtering."""
        search_query = {
            "query": {
                "bool": {
                    "must": [
                        # Exact match filters for memory isolation
                        *({"term": {f"payload.{k}": v}} for k, v in (filters or {}).items()),
                        # KNN vector search
                        {
                            "knn": {
                                "vector": {
                                    "vector": query,
                                    "k": limit
                                }
                            }
                        }
                    ]
                }
            }
        }

        response = self.client.search(index=self.collection_name, body=search_query)

        results = []
        for hit in response["hits"]["hits"]:
            results.append(
                OutputData(
                    id=hit["_id"],
                    score=hit["_score"],
                    payload=hit["_source"].get("payload", {})
                )
            )

        return results

    def delete(self, vector_id: str) -> None:
        """Delete a vector by ID."""
        self.client.delete(index=self.collection_name, id=vector_id)

    def update(self, vector_id: str, vector: Optional[List[float]] = None, payload: Optional[Dict] = None) -> None:
        """Update a vector and its payload."""
        doc = {}
        if vector is not None:
            doc["vector"] = vector
        if payload is not None:
            doc["payload"] = payload

        self.client.update(index=self.collection_name, id=vector_id, body={"doc": doc})

    def get(self, vector_id: str) -> Optional[OutputData]:
        """Retrieve a vector by ID."""
        try:
            response = self.client.get(index=self.collection_name, id=vector_id)
            return OutputData(
                id=response["_id"],
                score=1.0,  # Default score for direct get
                payload=response["_source"].get("payload", {}),
            )
        except KeyError as e:
            logger.warning(f"Missing key in Elasticsearch response: {e}")
            return None
        except TypeError as e:
            logger.warning(f"Invalid response type from Elasticsearch: {e}")
            return None
        except Exception as e:
            logger.error(f"Unexpected error while parsing Elasticsearch response: {e}")
            return None

    def list_cols(self) -> List[str]:
        """List all collections (indices)."""
        return list(self.client.indices.get_alias().keys())

    def delete_col(self) -> None:
        """Delete a collection (index)."""
        self.client.indices.delete(index=self.collection_name)

    def col_info(self, name: str) -> Any:
        """Get information about a collection (index)."""
        return self.client.indices.get(index=name)

    def list(self, filters: Optional[Dict] = None, limit: Optional[int] = None) -> List[List[OutputData]]:
        """List all memories."""
        query: Dict[str, Any] = {"query": {"match_all": {}}}

        if filters:
            query["query"] = {"bool": {"must": [{"match": {f"payload.{k}": v}} for k, v in filters.items()]}}

        if limit:
            query["size"] = limit

        response = self.client.search(index=self.collection_name, body=query)

        results = []
        for hit in response["hits"]["hits"]:
            results.append(
                OutputData(
                    id=hit["_id"],
                    score=1.0,  # Default score for list operation
                    payload=hit["_source"].get("payload", {}),
                )
            )

        return [results]

File: mem0-ts/src/oss/src/vector_stores/elasticsearch.ts
import type { Client } from "@elastic/elasticsearch";
import { VectorStore } from "./base";
import { SearchFilters, VectorStoreConfig, VectorStoreResult } from "../types";

interface ElasticsearchConfig extends VectorStoreConfig {
  /** Pre-configured Elasticsearch client instance (typed as `any` to keep the
   *  optional driver's types out of the published type declarations). */
  client?: any;
  host?: string;
  port?: number;
  cloudId?: string;
  apiKey?: string;
  username?: string;
  password?: string;
  collectionName: string;
  embeddingModelDims: number;
  dimension?: number;
  useSsl?: boolean;
  caCerts?: string;
  verifyCerts?: boolean;
  autoCreateIndex?: boolean;
  headers?: Record<string, string>;
}

const SAFE_FILTER_KEY = /^[a-zA-Z_][a-zA-Z0-9_]*$/;

// Mirrors the Python provider's _validate_filter. Filter keys are interpolated
// into `metadata.${key}` field paths, so reject anything that is not a plain
// identifier and reject non-scalar values before they reach the query.
function validateFilter(key: string, value: unknown): void {
  if (typeof key !== "string" || !SAFE_FILTER_KEY.test(key)) {
    throw new Error(`Invalid filter key: ${JSON.stringify(key)}`);
  }
  if (
    typeof value !== "string" &&
    typeof value !== "number" &&
    typeof value !== "boolean"
  ) {
    throw new Error(
      `Filter value for ${JSON.stringify(key)} must be string, number, or boolean`,
    );
  }
}

export class ElasticsearchDB implements VectorStore {
  private client!: Client;
  private readonly config: ElasticsearchConfig;
  private readonly collectionName: string;
  private readonly dimension: number;
  private readonly autoCreateIndex: boolean;
  private _initPromise?: Promise<void>;

  constructor(config: ElasticsearchConfig) {
    this.config = config;
    this.collectionName = config.collectionName;
    this.dimension = config.dimension || config.embeddingModelDims || 1536;
    this.autoCreateIndex = config.autoCreateIndex !== false;

    this.initialize().catch(console.error);
  }

  // The client is created lazily on first initialize() so the optional
  // `@elastic/elasticsearch` peer is only loaded when the store is actually used.
  private async ensureClient(): Promise<void> {
    if (this.client) return;

    const config = this.config;
    if (config.client) {
      this.client = config.client;
    } else {
      const params: Record<string, any> = {};

      if (config.cloudId) {
        params.cloud = { id: config.cloudId };
        if (config.apiKey) {
          params.auth = { apiKey: config.apiKey };
        }
      } else {
        const host = config.host || "localhost";
        const port = config.port || 9200;
        params.node = `${config.useSsl !== false ? "https" : "http"}://${host}:${port}`;

        if (config.apiKey) {
          params.auth = { apiKey: config.apiKey };
        } else if (config.username && config.password) {
          params.auth = {
            username: config.username,
            password: config.password,
          };
        }
      }

      if (config.verifyCerts !== undefined) {
        params.tls = { rejectUnauthorized: config.verifyCerts };
      }
      if (config.caCerts) {
        params.tls = { ...params.tls, ca: config.caCerts };
      }
      if (config.headers) {
        params.headers = config.headers;
      }

      let sdk: any;
      try {
        sdk = await import("@elastic/elasticsearch");
      } catch {
        throw new Error(
          "The '@elastic/elasticsearch' package is required to use the Elasticsearch vector store. Install it with: npm install @elastic/elasticsearch",
        );
      }

      this.client = new sdk.Client(params);
    }
  }

  async initialize(): Promise<void> {
    if (!this._initPromise) {
      this._initPromise = this._doInitialize();
    }
    return this._initPromise;
  }

  private async _doInitialize(): Promise<void> {
    await this.ensureClient();
    try {
      if (this.autoCreateIndex) {
        await this.ensureIndex(this.collectionName, this.dimension);
      }
      await this.ensureIndex("memory_migrations", 1);
    } catch (error) {
      console.error("Error initializing Elasticsearch:", error);
      throw error;
    }
  }

  private async ensureIndex(name: string, size: number): Promise<void> {
    const exists = await this.client.indices.exists({ index: name });
    if (!exists) {
      await this.client.indices.create({
        index: name,
        settings: {
          number_of_replicas: 1,
          number_of_shards: 5,
          refresh_interval: "1s",
        },
        mappings: {
          properties: {
            vector: {
              type: "dense_vector",
              dims: size,
              index: true,
              similarity: "cosine",
            },
            metadata: {
              type: "object",
              properties: {
                user_id: { type: "keyword" },
                agent_id: { type: "keyword" },
                run_id: { type: "keyword" },
              },
            },
          },
        },
      });
    }
  }

  async insert(
    vectors: number[][],
    ids: string[],
    payloads: Record<string, any>[],
  ): Promise<void> {
    await this.initialize();
    const operations: any[] = [];
    for (let i = 0; i < vectors.length; i++) {
      operations.push(
        { index: { _index: this.collectionName, _id: ids[i] } },
        { vector: vectors[i], metadata: payloads[i] || {} },
      );
    }

    await this.client.bulk({ operations });
  }

  async search(
    query: number[],
    topK: number = 5,
    filters?: SearchFilters,
  ): Promise<VectorStoreResult[]> {
    await this.initialize();
    const searchBody: Record<string, any> = {
      knn: {
        field: "vector",
        query_vector: query,
        k: topK,
        num_candidates: topK * 2,
      },
    };

    if (filters && Object.keys(filters).length > 0) {
      const filterConditions = Object.entries(filters).map(([key, value]) => {
        validateFilter(key, value);
        return { term: { [`metadata.${key}`]: value } };
      });
      searchBody.knn.filter = { bool: { must: filterConditions } };
    }

    const response = await this.client.search({
      index: this.collectionName,
      ...searchBody,
    });

    return response.hits.hits.map((hit: any) => ({
      id: hit._id,
      score: hit._score,
      payload: hit._source?.metadata || {},
    }));
  }

  async get(vectorId: string): Promise<VectorStoreResult | null> {
    await this.initialize();
    try {
      const response = await this.client.get({
        index: this.collectionName,
        id: vectorId,
      });
      return {
        id: response._id,
        payload: (response._source as any)?.metadata || {},
      };
    } catch (error: any) {
      if (error?.meta?.statusCode === 404) return null;
      throw error;
    }
  }

  async update(
    vectorId: string,
    vector: number[],
    payload: Record<string, any>,
  ): Promise<void> {
    await this.initialize();
    const doc: Record<string, any> = {};
    if (vector) doc.vector = vector;
    if (payload) doc.metadata = payload;

    await this.client.update({
      index: this.collectionName,
      id: vectorId,
      doc,
    });
  }

  async delete(vectorId: string): Promise<void> {
    await this.initialize();
    await this.client.delete({
      index: this.collectionName,
      id: vectorId,
    });
  }

  async deleteCol(): Promise<void> {
    await this.initialize();
    await this.client.indices.delete({ index: this.collectionName });
  }

  async list(
    filters?: SearchFilters,
    topK: number = 100,
  ): Promise<[VectorStoreResult[], number]> {
    await this.initialize();
    const query: Record<string, any> = { query: { match_all: {} } };

    if (filters && Object.keys(filters).length > 0) {
      const filterConditions = Object.entries(filters).map(([key, value]) => {
        validateFilter(key, value);
        return { term: { [`metadata.${key}`]: value } };
      });
      query.query = { bool: { must: filterConditions } };
    }

    query.size = topK;

    const response = await this.client.search({
      index: this.collectionName,
      ...query,
    });

    const results = response.hits.hits.map((hit: any) => ({
      id: hit._id,
      payload: hit._source?.metadata || {},
    }));

    return [results, results.length];
  }

  private generateUUID(): string {
    return "xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx".replace(/[xy]/g, (c) => {
      const r = (Math.random() * 16) | 0;
      const v = c === "x" ? r : (r & 0x3) | 0x8;
      return v.toString(16);
    });
  }

  async getUserId(): Promise<string> {
    await this.initialize();
    try {
      const response = await this.client.search({
        index: "memory_migrations",
        query: { match_all: {} },
        size: 1,
      });

      if (response.hits.hits.length > 0) {
        return (response.hits.hits[0]._source as any)?.metadata
          ?.user_id as string;
      }

      const randomUserId =
        Math.random().toString(36).substring(2, 15) +
        Math.random().toString(36).substring(2, 15);

      // memory_migrations is only ever read via match_all, never by vector,
      // so store no vector. A zero vector would be rejected by the cosine
      // dense_vector mapping ("does not support vectors with zero magnitude").
      await this.client.index({
        index: "memory_migrations",
        id: this.generateUUID(),
        document: { metadata: { user_id: randomUserId } },
      });

      return randomUserId;
    } catch (error) {
      console.error("Error getting user ID:", error);
      throw error;
    }
  }

  async setUserId(userId: string): Promise<void> {
    await this.initialize();
    try {
      const response = await this.client.search({
        index: "memory_migrations",
        query: { match_all: {} },
        size: 1,
      });

      const docId =
        response.hits.hits.length > 0
          ? response.hits.hits[0]._id
          : this.generateUUID();

      await this.client.index({
        index: "memory_migrations",
        id: docId,
        document: { metadata: { user_id: userId } },
      });
    } catch (error) {
      console.error("Error setting user ID:", error);
      throw error;
    }
  }
}
